### ISSUE TITLE
detection problem

### ISSUE THREAD

**Issue by sctrueew** [ultralytics/yolov5#5833]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi everyone,

I've trained a model with yolov5m and I have a problem with some classes which are linear. for example:
![label](https://user-images.githubusercontent.com/5433346/144085648-a43fd103-0d33-45cd-8a3c-415d74bd4c0c.jpg)  =>  ![pred](https://user-images.githubusercontent.com/5433346/144085655-eca00a00-e157-42c8-8bda-5c55ba711698.jpg)

Almost detection is 50% for the orange class but above 90% for the other classes.

Where is my problem?
![image](https://user-images.githubusercontent.com/5433346/144086932-f0a79b24-8828-484c-9993-3148dc5daa7f.png)

Thanks in advance

**Comment by glenn-jocher** [comment]
@sctrueew 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/](https://github.com/ultralytics/yolov5/issues/2844#issuecomment-851338384)

**Comment by Zengyf-CVer** [comment]
@sctrueew 
The number of instances in your orange category is relatively small.

**Comment by sctrueew** [comment]
@Zengyf-CVer Hi,

No, I have many instances about orange class than other classes.

number of instances for all classes:

**classid_numOfInstance**
1008_2911    **<==========> orange class** 
1009_633
1010_2741
1011_4976
1012_701
1013_136
1014_267
1015_2
1016_546
1017_340
1018_672
1019_663
1020_745
1022_628
1023_198
1024_137
1025_35
1026_13
1027_6
1028_56
1029_329
1030_488
1031_2092
1032_1

**Comment by Zengyf-CVer** [comment]
@sctrueew 
In the training set and validation set, does the number of instances of your orange class meet 8:2?

**Comment by sctrueew** [comment]
@Zengyf-CVer 

I used the training set for the valid set. I have 2911 instances for the orange class.

![image](https://user-images.githubusercontent.com/5433346/144186065-8d00f60e-2044-4d74-8ebb-30bcc545ba39.png)

two of the classes are strange. Did overfitting occur?

**Comment by Zengyf-CVer** [comment]
@sctrueew 
You'd better separate, 7:3 or 8:2 or 9:1. Except for the normal operation of the test program, do not use 1:1 for experiments.

**Comment by sctrueew** [comment]
@sctrueew Thanks.
I always used 1:1 for yolov4 and I thought that I can use it for this repo.

**Comment by Zengyf-CVer** [comment]
@sctrueew 
You can try the 8:2 instance division first to see if it works.

**Comment by sctrueew** [comment]
@Zengyf-CVer Hi,

I've trained again and my result is:

 Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 11/11 [00:09<00:00,  1.19it/s]                                                   
   all         500          2926         0.877     0.89      0.921        0.7
  Orange  500           254          0.877     0.472    0.573      0.394

**Comment by sctrueew** [comment]
@glenn-jocher Hi,

Could you please tell me, where the problem is?

Thanks

**Comment by glenn-jocher** [comment]
@sctrueew I have nothing to add beyond https://github.com/ultralytics/yolov5/issues/5833#issuecomment-982942590
